Add Hansen's Standardized Skewed T Distribution

chm-von-tla opened the pull request.

chm-von-tla commented:
The first commit is a minor style fix
The motivation for the second commit is discussed in #87

s-broda commented on src/univariatestandardizeddistributions.jl:
```diff
@@ -253,3 +253,78 @@ function quantile(dist::StdGED, q::Real)
+Create a standardized skewed t distribution with `v` degrees of freedom and `λ` shape parameter. `ν,λ`` can be passed
+as scalars or vectors.
+"""
+StdSkewT(ν,λ) = StdSkewT([ν λ])
```
```StdSkewT(ν,λ) = StdSkewT([ν, λ])```

s-broda commented on src/univariatestandardizeddistributions.jl:
```diff
@@ -253,3 +253,78 @@ function quantile(dist::StdGED, q::Real)
+as scalars or vectors.
+"""
+StdSkewT(ν,λ) = StdSkewT([ν λ])
+StdSkewT(ν::Integer,λ::Integer) = StdSkewT(float(ν),float(λ))
```
This can be removed. That conversion happens automatically.

chm-von-tla replied:
Actually, it doesn't

s-broda commented on src/univariatestandardizeddistributions.jl:
```diff
@@ -253,3 +253,78 @@ function quantile(dist::StdGED, q::Real)
+"""
+StdSkewT(ν,λ) = StdSkewT([ν λ])
+StdSkewT(ν::Integer,λ::Integer) = StdSkewT(float(ν),float(λ))
+StdSkewT(v::Vector{T},λ::Vector{T}) where {T} = StdSkewT{T}(v,λ)
```
This doesn't work because there is no such inner constructor. Also, it probably shouldn't work, so just remove it.

s-broda commented on src/univariatestandardizeddistributions.jl:
```diff
@@ -253,3 +253,78 @@ function quantile(dist::StdGED, q::Real)
+StdSkewT(v::Vector{T},λ::Vector{T}) where {T} = StdSkewT{T}(v,λ)
+StdSkewT(coefs::Vector{T}) where {T} = StdSkewT{T}(coefs)
+
+(rand(d::StdSkewT{T})::T) where {T}  =  (ν=d.coefs[1];λ=d.coefs[2]; quantile(d,rand(1)))
```
`rand(d::StdSkewT{T}) where {T}  =  (ν=d.coefs[1];λ=d.coefs[2]; quantile(d, rand()))`

chm-von-tla replied:
Is this correct? I see ν and λ being assigned and then not used

s-broda commented on src/univariatestandardizeddistributions.jl:
```diff
@@ -253,3 +253,78 @@ function quantile(dist::StdGED, q::Real)
+function quantile(d::StdSkewT{T}, q::T) where T
+    ν = d.coefs[1]
+    λ = d.coefs[2]
+    xa = a(d,d.coefs)
```
`xa = a(typeof(d), d.coefs)`

s-broda commented on src/univariatestandardizeddistributions.jl:
```diff
@@ -253,3 +253,78 @@ function quantile(dist::StdGED, q::Real)
+    ν = d.coefs[1]
+    λ = d.coefs[2]
+    xa = a(d,d.coefs)
+    xb = b(d,d.coefs)
```
xb = b(typeof(d), d.coefs)

chm-von-tla commented:
I incorporated the changes you requested, added tests (by basically imitating the tests for the other distributions), I added a mention of the distribution in the docs (I'm not familiar with the process so I may have made mistakes) and optimized the performance (see my last post on #87, and/or commit message for 49e7858)

s-broda commented:
This is starting to look good. There's just a doctest that's still failing: https://travis-ci.org/github/s-broda/ARCHModels.jl/jobs/762500760

Do you know how to address this? You basically need to update the output in src/usage.md, lines 331-362. If not, I can merge it as is and fix it myself.

Cheers
Simon

chm-von-tla commented:
No, I am not familiar with what's going on in the lines you mentioned. However I also spotted a different thing on those logs: after the error on line 990 it is complaining that the evaluated output differs from what is expected (there is a space missing in the function call)

chm-von-tla commented:
It would be better if you fixed it, I haven't used julia's documentation system before

s-broda commented:
OK, I'll merge and fix it myself.

Thanks for your contribution!

chm-von-tla commented:
Thank you for the advice along the way!
